---
title: "Ailaoshan_preOTUs_networks.Rmd"
author: "Chris Baker"
email: "bakerccm@gmail.com"
date: "16/07/2019"
output: html_document
---

This file takes the already-processed preOTU data from Ailaoshan_preOTUs.Rdata and makes network plots. These plots will be used by Doug to assign final taxonomies to the preOTUs. PreOTUs with the same final taxonomic assignment will be collapsed to generate the OTU dataset that we will analyse using occupancy methods.

```{r setup, include=FALSE}
library("tidyverse")
library("igraph")
library("Hmisc") # for rcorr
```

```{r clear old variables, include=FALSE}
rm(list=ls())
```

```{r read in preOTU table data, include=FALSE}
# generated with Ailaoshan_OTU_tables_preOTUs.Rmd
load(file="Ailaoshan_preOTUs.Rdata")
```

```{r calculate pairwise Spearman correlations within datasets}

reads.tbl <- list()
reads.cor <- list()
reads.cor.r <- list()
reads.cor.P <- list()

for (i in c("LSU", "SSU")) {
    # format reads as OTU table
        reads.tbl[[i]] <- read.tidy[[i]] %>% spread(key = "OTU", value = "reads") %>% rename_at(vars(-"Lab_ID"), .funs = function (X) paste0(i, ".", X))
    # calculate correlations based on presence/absence
        reads.cor[[i]] <- rcorr(reads.tbl[[i]] %>% select(starts_with(i)) %>% mutate_all(.funs = function (X) ifelse(X > 0, 1, 0)) %>% as.matrix(), type="spearman")
        reads.cor.r[[i]] <- reads.cor[[i]]$r
        reads.cor.P[[i]] <- reads.cor[[i]]$P
    # simplify OTU numbers
        rownames(reads.cor.r[[i]]) <- rownames(reads.cor.r[[i]]) %>% gsub(paste0(i, "."), "", .)
        colnames(reads.cor.r[[i]]) <- colnames(reads.cor.r[[i]]) %>% gsub(paste0(i, "."), "", .)
    # remove NAs
        reads.cor.r[[i]][is.na(reads.cor.r[[i]])] <- 0 # there aren't any of these
    # remove negative correlations
        reads.cor.r[[i]][reads.cor.r[[i]] < 0] <- 0
    # remove nonsignificant correlations
    # after FDR correction
        reads.cor.r[[i]][p.adjust(reads.cor.P[[i]], "fdr") > 0.05] <- 0
}
```

```{r calculate pairwise Spearman correlations between datasets}

# make joint read table, removing any OTUs with no reads in the overlapping samples

reads.table.all <- inner_join(reads.tbl$LSU, reads.tbl$SSU, by = "Lab_ID") %>%
    gather(-Lab_ID, key = "OTU", value = "reads") %>%
    group_by(OTU) %>% filter(sum(reads) > 0) %>% ungroup() %>%
    spread(key = "OTU", value = "reads") %>%
    mutate_all(.funs = function (X) ifelse(X > 0, 1, 0))

crosscor <- rcorr(reads.table.all %>% select(-Lab_ID) %>% mutate_all(.funs = function (X) ifelse(X > 0, 1, 0)) %>% as.matrix(), type="spearman")

crosscor.r <- crosscor$r[grepl("LSU.", rownames(crosscor$r)), grepl("SSU.", colnames(crosscor$r))]
rownames(crosscor.r) <- rownames(crosscor.r) %>% gsub("LSU.", "", .)  # each row is an OTU in the LSU dataset
colnames(crosscor.r) <- colnames(crosscor.r) %>% gsub("SSU.", "", .)  # each col is an OTU in the SSU dataset

crosscor.P <- crosscor$P[grepl("LSU.", rownames(crosscor$P)), grepl("SSU.", colnames(crosscor$P))]
rownames(crosscor.P) <- rownames(crosscor.P) %>% gsub("LSU.", "", .)  # each row is an OTU in the LSU dataset
colnames(crosscor.P) <- colnames(crosscor.P) %>% gsub("SSU.", "", .)  # each col is an OTU in the SSU dataset

# remove NAs
    crosscor.r[is.na(crosscor.r)] <- 0 # there aren't any of these
# remove negative correlations
    crosscor.r[crosscor.r < 0] <- 0
# remove nonsignificant correlations
# after FDR correction
    crosscor.r[p.adjust(crosscor.P, "fdr") > 0.05] <- 0
```

```{r add best guess taxonomy information to read tables for each dataset}
for (i in c("LSU", "SSU")) {
    # split concatenated taxon strings
        taxon.tbl[[i]] <- taxon.tbl[[i]] %>%
            mutate(order = ifelse(order == "root,unk", "unk", order)) %>%
            separate(family, into = c("family.order", "family.family"), remove = FALSE) %>%
            separate(genus, into = c("genus.order", "genus.family", "genus.genus"), remove = FALSE) %>%
            separate(species, into = c("species.order", "species.family", "species.genus", "species.species"), remove = FALSE)
    # assemble 'best guess' taxon for each preOTU
    # start with species; if taxon is unk or NA then go to next higher taxon assignment, otherwise keep it
    # note that this ignores protax probabilities and any inconsistencies between taxon assignments at different levels.
    # note also that this is distinct from the besthit generated by protax
        taxon.tbl[[i]] <- taxon.tbl[[i]] %>%
            mutate(best.guess = species) %>%
            mutate(best.guess = ifelse(species.species == "unk" | is.na(species), ifelse(genus.genus == "unk" | is.na(genus.genus), ifelse(family.family == "unk", order, family), genus), best.guess)) %>%
            mutate(best.guess = ifelse(best.guess == "unk", NA, best.guess))
    # get probability info for whichever taxon level was chosen as the 'best guess'
        taxon.tbl[[i]] <- taxon.tbl[[i]] %>%
            mutate(best.guess.prob = NA) %>%
            mutate(best.guess.prob = ifelse(!is.na(best.guess) & best.guess == species, prob_species, best.guess.prob)) %>%
            mutate(best.guess.prob = ifelse(!is.na(best.guess) & best.guess == genus, prob_genus, best.guess.prob)) %>%
            mutate(best.guess.prob = ifelse(!is.na(best.guess) & best.guess == family, prob_family, best.guess.prob)) %>%
            mutate(best.guess.prob = ifelse(!is.na(best.guess) & best.guess == order, prob_order, best.guess.prob)) %>%
            mutate(best.guess.prob = sprintf("%.2f", best.guess.prob))
}
```

```{r makes graphs}
OTU.graph <- list()
# graphs for each dataset
    for (i in c("LSU", "SSU")) {
        OTU.graph[[i]] <- graph.adjacency(reads.cor.r[[i]], weighted = T, mode="undirected", diag=F)  # reads.cor$xxx is used as the adjacency matrix for this graph
        V(OTU.graph[[i]])$OTU.no <- V(OTU.graph[[i]])$name
    }
# bipartite graph for the two datasets together
    OTU.graph$all <- graph.incidence(crosscor.r, weighted = T)  # reads.cor$xxx is used as the incidence matrix for this graph
    V(OTU.graph$all)$type[V(OTU.graph$all)$type == FALSE] <- "LSU"
    V(OTU.graph$all)$type[V(OTU.graph$all)$type == TRUE] <- "SSU"
    V(OTU.graph$all)$OTU.no <- V(OTU.graph$all)$name
```

```{r set some graphical parameters}
for (i in names(OTU.graph)) {
    V(OTU.graph[[i]])$label.family <- "Helvetica"
    V(OTU.graph[[i]])$size <- 6
    V(OTU.graph[[i]])$shape <- "circle"
    V(OTU.graph[[i]])$frame.color <- NA
    V(OTU.graph[[i]])$label.cex <- 0.7
    V(OTU.graph[[i]])$label.degree <- 0
    #V(OTU.graph[[i]])$label.dist <- 1
    #V(OTU.graph[[i]])$label.color <- V(OTU.graph)$color
    V(OTU.graph[[i]])$label.color <- "black"
    E(OTU.graph[[i]])$width <- E(OTU.graph[[i]])$weight * 10
}
# color nodes by dataset: LSU/16S = blue, SSU/12S = red
    V(OTU.graph$LSU)$color <- "deepskyblue"
    V(OTU.graph$SSU)$color <- "red"
    V(OTU.graph$all)$color <- ifelse(V(OTU.graph$all)$type == "LSU", "deepskyblue", "red")
```

```{r add best guess taxon data to vertices}
# LSU, SSU graphs

    for (i in c("LSU", "SSU")) {
        taxon.info <- taxon.tbl[[i]] %>% filter(protaxmod == "weighted")
        for (j in names(taxon.tbl[[i]])[names(taxon.tbl[[i]]) != "queryID"]) {
            OTU.graph[[i]] <- OTU.graph[[i]] %>% set_vertex_attr(name = j, value = taxon.info[[j]][match(V(OTU.graph[[i]])$OTU.no, taxon.info$queryID, 0)])
        }
        rm(taxon.info)
    }

# bipartite graph for LSU+SSU

    V(OTU.graph$all)$type.OTUno <- paste0(V(OTU.graph$all)$type, ".", V(OTU.graph$all)$OTU.no)
    taxon.tbl$all <- bind_rows(LSU = taxon.tbl$LSU, SSU = taxon.tbl$SSU, .id = "dataset")
    taxon.tbl$all <- taxon.tbl$all %>% mutate(type.OTUno = paste0(dataset, ".", queryID))

    taxon.info <- taxon.tbl$all %>% filter(protaxmod == "weighted")
    for (j in names(taxon.tbl$all)[!(names(taxon.tbl$all) %in% c("queryID"))]) {
        OTU.graph$all <- OTU.graph$all %>%
            set_vertex_attr(name = j, value = taxon.info[[j]][match(V(OTU.graph$all)$type.OTUno, taxon.info$type.OTUno, 0)])
    }
    rm(taxon.info)
    
# set taxon names as vertex labels
for (i in names(OTU.graph)) {
    V(OTU.graph[[i]])$name <- paste(V(OTU.graph[[i]])$OTU.no, V(OTU.graph[[i]])$best.guess, V(OTU.graph[[i]])$best.guess.prob, sep="\n")
}
```

```{r color edges black if taxon best guesses match}
for (i in names(OTU.graph)) {
    E(OTU.graph[[i]])$color <- ifelse(head_of(OTU.graph[[i]],E(OTU.graph[[i]]))$best.guess == tail_of(OTU.graph[[i]],E(OTU.graph[[i]]))$best.guess, "black", "lightgrey")
}
```

```{r set threshold for edges}
for (i in names(OTU.graph)) {
    # remove edges with Spearman correlation < 0.1
    # note that edges with negative correlations or non-significant correlations (after FDR correction)
    # are already set to zero weight by code above so this removes those edges
    OTU.graph[[i]] <- delete_edges(OTU.graph[[i]], which(E(OTU.graph[[i]])$weight < 0.1))
    # uncomment this to actually delete orphaned preOTUs, otherwise leave them in the graph
    # OTU.graph[[i]] <- delete_vertices(OTU.graph[[i]], degree(OTU.graph[[i]]) == 0)
}
```

```{r size vertices by (raw) occupancy}
occupancy <- list()

# LSU, SSU graphs

    for (i in c("LSU", "SSU")) {
        occupancy[[i]] <- read.tidy[[i]] %>%
            group_by(OTU) %>%
            dplyr::summarize(occupancy = sum(reads > 0)/n())
        V(OTU.graph[[i]])$occupancy <- occupancy[[i]]$occupancy[match(V(OTU.graph[[i]])$OTU.no, occupancy[[i]]$OTU, 0)]
        V(OTU.graph[[i]])$size <- sqrt(V(OTU.graph[[i]])$occupancy) *20
    }

# bipartite graph for LSU+SSU

    occupancy$all <- bind_rows(LSU = occupancy$LSU, SSU = occupancy$SSU, .id = "dataset") %>%
        mutate(dataset.OTU = paste0(dataset, ".", OTU))
    V(OTU.graph$all)$occupancy <- occupancy$all$occupancy[match(V(OTU.graph$all)$type.OTUno, occupancy$all$dataset.OTU, 0)]
    V(OTU.graph$all)$size <- sqrt(V(OTU.graph$all)$occupancy) * 20

rm(occupancy)
```

```{r draw plots}

    pdf("Ailaoshan_preOTUs_networks/preOTU_network_LSU_weighted.pdf", width=20, height=20)
        
        set.seed(1)
        plot(OTU.graph$LSU, main="LSU taxa (weighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$LSU, V(OTU.graph$LSU)$class.inferred != "Mammals" | is.na(V(OTU.graph$LSU)$class.inferred) == TRUE),
             main="LSU taxa: Mammals only (weighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$LSU, V(OTU.graph$LSU)$class.inferred != "Birds" | is.na(V(OTU.graph$LSU)$class.inferred) == TRUE),
             main="LSU taxa: Birds only (weighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$LSU, V(OTU.graph$LSU)$class.inferred != "Amphibians" | is.na(V(OTU.graph$LSU)$class.inferred) == TRUE),
             main="LSU taxa: Amphibians only (weighted)")
        
    dev.off()

    pdf("Ailaoshan_preOTUs_networks/preOTU_network_SSU_weighted.pdf", width=20, height=20)
    
        set.seed(1)
        plot(OTU.graph$SSU, main="SSU taxa (weighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$SSU, V(OTU.graph$SSU)$class.inferred != "Mammals" | is.na(V(OTU.graph$SSU)$class.inferred) == TRUE),
             main="SSU taxa: Mammals only (weighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$SSU, V(OTU.graph$SSU)$class.inferred != "Birds" | is.na(V(OTU.graph$SSU)$class.inferred) == TRUE),
             main="SSU taxa: Birds only (weighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$SSU, V(OTU.graph$SSU)$class.inferred != "Amphibians" | is.na(V(OTU.graph$SSU)$class.inferred) == TRUE),
             main="SSU taxa: Amphibians only (weighted)")
        
    dev.off()

    pdf("Ailaoshan_preOTUs_networks/preOTU_network_bipartite_weighted.pdf", width=20, height=20)
        
        set.seed(1)
        plot(OTU.graph$all, main="all taxa (weighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$all, V(OTU.graph$all)$class.inferred != "Mammals" | is.na(V(OTU.graph$all)$class.inferred) == TRUE),
             main="all taxa: Mammals only (weighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$all, V(OTU.graph$all)$class.inferred != "Birds" | is.na(V(OTU.graph$all)$class.inferred) == TRUE),
             main="all taxa: Birds only (weighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$all, V(OTU.graph$all)$class.inferred != "Amphibians" | is.na(V(OTU.graph$all)$class.inferred) == TRUE),
             main="all taxa: Amphibians only (weighted)")
        
    dev.off()
    
```

```{r redraw graphs with unweighted taxa}
# add unweighted taxon info to vertices

    # LSU, SSU graphs
        for (i in c("LSU", "SSU")) {
            taxon.info <- taxon.tbl[[i]] %>% filter(protaxmod == "unweighted")
            for (j in names(taxon.tbl[[i]])[names(taxon.tbl[[i]]) != "queryID"]) {
                OTU.graph[[i]] <- OTU.graph[[i]] %>%
                    set_vertex_attr(name = j, value = taxon.info[[j]][match(V(OTU.graph[[i]])$OTU.no, taxon.info$queryID, 0)])
            }
            rm(taxon.info)
        }

    # bipartite graph for LSU+SSU
        taxon.info <- taxon.tbl$all %>% filter(protaxmod == "unweighted")
        for (j in names(taxon.tbl$all)[names(taxon.tbl$all) != "queryID"]) {
            OTU.graph$all <- OTU.graph$all %>%
                set_vertex_attr(name = j, value = taxon.info[[j]][match(V(OTU.graph$all)$type.OTUno, taxon.info$type.OTUno, 0)])
        }
        rm(taxon.info)

# color edges black if taxon best guess matches

    for (i in names(OTU.graph)) {
        E(OTU.graph[[i]])$color <- ifelse(head_of(OTU.graph[[i]],E(OTU.graph[[i]]))$best.guess == tail_of(OTU.graph[[i]],E(OTU.graph[[i]]))$best.guess, "black", "lightgrey")
    }

# set taxon names as vertex labels

    for (i in names(OTU.graph)) {
        V(OTU.graph[[i]])$name <- paste(V(OTU.graph[[i]])$OTU.no, V(OTU.graph[[i]])$best.guess, V(OTU.graph[[i]])$best.guess.prob, sep="\n")
    }

# draw plots

    pdf("Ailaoshan_preOTUs_networks/preOTU_network_LSU_unweighted.pdf", width=20, height=20)
        
        set.seed(1)
        plot(OTU.graph$LSU, main="LSU taxa (unweighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$LSU, V(OTU.graph$LSU)$class.inferred != "Mammals" | is.na(V(OTU.graph$LSU)$class.inferred) == TRUE),
             main="LSU taxa: Mammals only (unweighted)")

        set.seed(1)
        plot(delete_vertices(OTU.graph$LSU, V(OTU.graph$LSU)$class.inferred != "Birds" | is.na(V(OTU.graph$LSU)$class.inferred) == TRUE),
             main="LSU taxa: Birds only (unweighted)")

        set.seed(1)
        plot(delete_vertices(OTU.graph$LSU, V(OTU.graph$LSU)$class.inferred != "Amphibians" | is.na(V(OTU.graph$LSU)$class.inferred) == TRUE),
             main="LSU taxa: Amphibians only (unweighted)")

    dev.off()

    pdf("Ailaoshan_preOTUs_networks/preOTU_network_SSU_unweighted.pdf", width=20, height=20)

        set.seed(1)
        plot(OTU.graph$SSU, main="SSU taxa (unweighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$SSU, V(OTU.graph$SSU)$class.inferred != "Mammals" | is.na(V(OTU.graph$SSU)$class.inferred) == TRUE),
             main="SSU taxa: Mammals only (unweighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$SSU, V(OTU.graph$SSU)$class.inferred != "Birds" | is.na(V(OTU.graph$SSU)$class.inferred) == TRUE),
             main="SSU taxa: Birds only (unweighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$SSU, V(OTU.graph$SSU)$class.inferred != "Amphibians" | is.na(V(OTU.graph$SSU)$class.inferred) == TRUE),
             main="SSU taxa: Amphibians only (unweighted)")
    
    dev.off()

    pdf("Ailaoshan_preOTUs_networks/preOTU_network_bipartite_unweighted.pdf", width=20, height=20)
        
        set.seed(1)
        plot(OTU.graph$all, main="all taxa (unweighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$all, V(OTU.graph$all)$class.inferred != "Mammals" | is.na(V(OTU.graph$all)$class.inferred) == TRUE),
             main="all taxa: Mammals only (unweighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$all, V(OTU.graph$all)$class.inferred != "Birds" | is.na(V(OTU.graph$all)$class.inferred) == TRUE),
             main="all taxa: Birds only (unweighted)")
        
        set.seed(1)
        plot(delete_vertices(OTU.graph$all, V(OTU.graph$all)$class.inferred != "Amphibians" | is.na(V(OTU.graph$all)$class.inferred) == TRUE),
             main="all taxa: Amphibians only (unweighted)")
    
    dev.off()
```

```{r}

```
