---
title: "Ailaoshan_occupancy_MCMC_prepare.Rmd"
author: "Chris Baker"
email: "bakerccm@gmail.com"
date: "25/10/2019"
output: html_document
---

```{r setup}
library(tidyverse)
```

```{r clear old variables}
rm(list=ls())
```

```{r read in raw data}
# generated with Ailaoshan_OTU_table.Rmd
    load(file="../../Ailaoshan_OTU_table.Rdata")
```

```{r augment read data and convert to presence-absence}
leech.augmented <- list()

for (i in c("LSU", "SSU")) {
    leech.augmented[[i]] <- leech %>%
        # add extra rows for Polygon_IDs with missing data
            bind_rows(leech.supplement) %>%
        # split by dataset
            filter(dataset == i) %>%
        # convert reads to presence-absence
            mutate(present = as.integer(ifelse(reads > 0, 1, 0))) %>%
            select(-reads) %>%
        # add counts of other taxa
            group_by(Polygon_ID, Lab_ID) %>%
            mutate(othertaxa = sum(present) - present) %>%
            ungroup() %>%
        # scale variables
            mutate(othertaxa_scaled = scale(othertaxa),
                leech_qty_scaled = scale(leech_qty),
                fraction.reads.humans_scaled = scale(fraction.reads.humans)) %>%
            select(-othertaxa, -leech_qty, -fraction.reads.humans)
}
rm(i)
```

```{r make empty lists of arrays to be filled with data}

# (see K&R p.644)

# get dimensions for data arrays
    nsites <- list() # number of sites i.e. Polygon_IDs for each dataset
    nreps <- list()  # (maximum) number of replicates per Polygon_ID in each dataset
    nspec <- list()  # total number of observed species for each dataset

# make lists to be filled with data arrays
    y <- list()             # species presences for each site, to be indexed as [nsites, nreps, nspec]
    othertaxa <- list()     # number of other taxa detected, to be indexed as [nsites, nreps, nspec]
                            # (so that each taxon is excluded from the predictor for its own detection probability)
    numleeches <- list()    # number of leeches per replicate, to be indexed as [nsites, nreps]
    humanfraction <- list() # fraction of reads in each replicate from humans, to be indexed as [nsites, nreps]
    
# make empty data arrays
    for (i in names(leech.augmented)) {
        #get max dimensions for sizing data arrays
            nsites[[i]] <- leech.augmented[[i]] %>%
                select(Polygon_ID) %>% distinct() %>% nrow()
            nreps[[i]] <- leech.augmented[[i]] %>%
                group_by(Polygon_ID, OTU) %>% tally() %>%
                group_by(Polygon_ID) %>% summarise(max.n = max(n)) %>%
                ungroup() %>% pull(max.n) %>% max()
            nspec[[i]] <- leech.augmented[[i]] %>%
                select(OTU) %>% distinct() %>% nrow()
        # fill lists with arrays
            # 3D arrays indexed as [nsites, nreps, nspec]
                y[[i]] <- array(NA, dim=c(nsites[[i]], nreps[[i]], nspec[[i]]))
                othertaxa[[i]] <- array(NA, dim=c(nsites[[i]], nreps[[i]], nspec[[i]]))
            # 2D arrays indexed as [nsites, nreps]
                numleeches[[i]] <- array(NA, dim=c(nsites[[i]], nreps[[i]]))
                humanfraction[[i]] <- array(NA, dim=c(nsites[[i]], nreps[[i]]))
    }
        
# check data dimenions
# (note that nsites now includes Polygon_IDs that never had any reads but also extra Polygon_IDs
# derived from Ranger_IDs, so exceeds number of actual polygons
    unlist(nsites) #LSU: 209, SSU: 209
    unlist(nreps) #LSU: 40, SSU: 38
    unlist(nspec) #LSU: 59, SSU: 72

rm(i, nsites, nreps, nspec)
```

```{r prepare data}
OTU.labels <- list() # species
polygon.labels <- list()      # sites
replicate.labels <- list() # replicates

for (i in names(leech.augmented)) {

    # names for ensuring that arrays are aligned
        OTU.labels[[i]] <- leech.augmented[[i]] %>% select(OTU) %>% distinct() %>% arrange(OTU) %>% pull(OTU)
        polygon.labels[[i]] <- leech.augmented[[i]] %>% select(Polygon_ID) %>% distinct() %>% arrange(Polygon_ID) %>% pull(Polygon_ID)
        replicate.labels[[i]] <- leech.augmented[[i]] %>% select(replicate_no) %>% distinct() %>% arrange(replicate_no) %>% pull(replicate_no)

    # fill y[[i]] with data
        temp <- leech.augmented[[i]] %>% 
            select(Polygon_ID, OTU, replicate_no, present) %>%
            pivot_wider(names_from = replicate_no, values_from = present)
        for (k in seq_along(OTU.labels[[i]])) {
            # this is just the wide data from before
                temp.k <- temp %>% filter(OTU == OTU.labels[[i]][k]) # filter to the kth OTU
            # arrange rows and then remove Polygon_ID column
                temp.k <- temp.k[match(polygon.labels[[i]], temp.k$Polygon_ID),] %>% select(starts_with("replicate_"))
            # arrange columns
                temp.k <- temp.k[,match(replicate.labels[[i]], names(temp.k))]
            # convert to matrix
                y[[i]][,,k] <- temp.k %>% data.matrix()
        }
        rm(temp, k, temp.k)
        dimnames(y[[i]]) <- list(polygon.labels[[i]], replicate.labels[[i]], OTU.labels[[i]])

    # fill othertaxa[[i]] with data
        temp <- leech.augmented[[i]] %>% 
            select(Polygon_ID, OTU, replicate_no, othertaxa_scaled) %>%
            pivot_wider(names_from = replicate_no, values_from = othertaxa_scaled)
        for (k in seq_along(OTU.labels[[i]])) {
            # this is just the wide data from before
                temp.k <- temp %>% filter(OTU == OTU.labels[[i]][k]) # filter to the kth OTU
            # arrange rows and then remove Polygon_ID column
                temp.k <- temp.k[match(polygon.labels[[i]], temp.k$Polygon_ID),] %>% select(starts_with("replicate_"))
            # arrange columns
                temp.k <- temp.k[,match(replicate.labels[[i]], names(temp.k))]
            # convert to matrix
                othertaxa[[i]][,,k] <- temp.k %>% data.matrix()
        }
        rm(temp, k, temp.k)
        dimnames(y[[i]]) <- list(polygon.labels[[i]], replicate.labels[[i]], OTU.labels[[i]])

    # fill numleeches[[i]] with data
        temp <- leech.augmented[[i]] %>%
            select(dataset, Polygon_ID, replicate_no, leech_qty_scaled) %>%
            distinct() %>% pivot_wider(names_from = replicate_no, values_from = leech_qty_scaled)
        temp <- temp[match(polygon.labels[[i]], temp$Polygon_ID),]
        temp <- temp %>% select(starts_with("replicate_"))
        temp <- temp[,match(replicate.labels[[i]], names(temp))]
        numleeches[[i]] <- temp %>% data.matrix()
        rm(temp)
        dimnames(numleeches[[i]]) <- list(polygon.labels[[i]], replicate.labels[[i]])

    # fill humanfraction[[i]] with data
    # actually should we be including human reads for excluded samples?
        temp <- leech.augmented[[i]] %>%
            select(dataset, Polygon_ID, replicate_no, fraction.reads.humans_scaled) %>%
            distinct() %>% pivot_wider(names_from = replicate_no, values_from = fraction.reads.humans_scaled)
        temp <- temp[match(polygon.labels[[i]], temp$Polygon_ID),]
        temp <- temp %>% select(starts_with("replicate_"))
        temp <- temp[,match(replicate.labels[[i]], names(temp))]
        humanfraction[[i]] <- temp %>% data.matrix()
        rm(temp)
        dimnames(humanfraction[[i]]) <- list(polygon.labels[[i]], replicate.labels[[i]])
        
}

rm(i, OTU.labels, polygon.labels, replicate.labels)
```

```{r samples per Polygon_ID}
# frequency distribution for number of samples per Polygon_ID (i.e. per 'site')
nsurvey <- list()
for (i in names(leech.augmented)) {
    table(nsurvey[[i]] <- apply(y[[i]][,,1], 1, function (X) sum(!is.na(X))))
}
rm(i)
```

```{r observed number of occupied sites}
obs.occ <- list()
for (i in names(leech.augmented)) {
    suppressWarnings( # for the 'all NA' sites, max returns -Inf with a warning
        temp <- apply(y[[i]], c(1,3), max, na.rm = TRUE)
    )
    temp[temp == -Inf] <- NA # change -Inf to NA for any sites that were all NA
    sort(obs.occ[[i]] <- apply(temp, 2, sum, na.rm = TRUE), decreasing = TRUE)
    rm(temp)
}
rm(i)
```

```{r observed number of species per Polygon_ID, fig.height=2, fig.width=3, eval=FALSE}
#(i.e. per site)
# seems like this is best accomplished with tidy data rather than this code (from K&R)
par(mfrow = c(1,2))
for (i in names(leech.augmented)) {
    suppressWarnings( # for the 'all NA' sites, max returns -Inf with a warning
        temp <- apply(y[[i]], c(1,3), max, na.rm = TRUE)
    )
    temp[temp == -Inf] <- NA # change -Inf to NA for any sites that were all NA
    sort(C <- apply(temp, 1, sum), decreasing=TRUE)     # compute and print sorted species counts
    #print(table(C))
    plot(table(C), xlim = c(0, 15), ylim = c(0, 30), xlab = "Observed number of species", ylab = "Number of Polygon_IDs", frame = FALSE, xaxt='n', yaxt='n', main=i)
    axis(side=1, at=seq(0,15,5))
    axis(side=2, at=seq(0,30,10))
    abline(v = mean(C, na.rm = TRUE), col = "red", lwd = 1, lty=3)
    rm(temp, C)
}
rm(i)
```

```{r observed occurence values as starting values for z}
# note that these vary by dataset
zst <- list()

# now we have some site/species combinations that don't even have one replicate
#    any(is.na(y$LSU[,1,]))
#    any(is.na(y$SSU[,1,]))

for (i in names(leech.augmented)) {
    suppressWarnings( # for the 'all NA' sites, max returns -Inf with a warning
        zst[[i]] <- apply(y[[i]], c(1,3), max, na.rm = TRUE) # Observed occurrence as starting values for z
    )
    zst[[i]][!is.finite(zst[[i]])] <- 1 # performs replacement for values of -Inf (arising from sites with all NAs, i.e. extra Polygon_IDs without any replicates) as well as NA
}
rm(i)
```

```{r extract and scale per-site occupancy covariates}
site.covariates.scaled  <- list()
for (i in names(leech.augmented)) {
    site.covariates <- leech.augmented[[i]] %>%
        select(Polygon_ID, elevation_median, tpi_median, distance_to_road_median, distance_to_stream_median, distance_to_nature_reserve_boundary) %>%
        distinct() %>%
        right_join(tibble(Polygon_ID = y[[i]] %>% rownames()), by = "Polygon_ID")
    site.covariates.scaled[[i]] <- scale(site.covariates[,-1])
    rownames(site.covariates.scaled[[i]]) <- site.covariates$Polygon_ID
    rm(site.covariates)
}
rm(i)
```

```{r MCMC settings}
mcmc.settings <- list(
    ni = 40000,
    nt = 20,
    nb = 10000,
    nc = 3
)
```

```{r package data for JAGS}
model.data <- list()

model.data[["model01a"]] <- list()
model.data[["model01b"]] <- list()
model.data[["model01c"]] <- list()
for (i in names(leech.augmented)) {
    model.data[["model01a"]][[i]] <- list(y = y[[i]], nsite = dim(y[[i]])[1], nrep = dim(y[[i]])[2], nspec = dim(y[[i]])[3],
         elev = site.covariates.scaled[[i]][,"elevation_median"],
         tpi = site.covariates.scaled[[i]][,"tpi_median"],
         road = site.covariates.scaled[[i]][,"distance_to_road_median"],
         stream = site.covariates.scaled[[i]][,"distance_to_stream_median"],
         reserve = site.covariates.scaled[[i]][,"distance_to_nature_reserve_boundary"],
         numleeches = numleeches[[i]],
         R = matrix(c(5,0,0,1), ncol = 2), df = 3,
         zst = zst[[i]]
    )
    # copy over most data
        model.data[["model01b"]][[i]] <- model.data[["model01a"]][[i]]
        model.data[["model01c"]][[i]] <- model.data[["model01a"]][[i]]
    # add sampling covariates
        model.data[["model01b"]][[i]]$othertaxa <- othertaxa[[i]]
        model.data[["model01c"]][[i]]$humanfraction <- humanfraction[[i]]
}

model.data[["model12a"]] <- list()
model.data[["model12b"]] <- list()
model.data[["model12c"]] <- list()
for (i in names(leech.augmented)) {
    model.data[["model12a"]][[i]] <- list(y = y[[i]], nsite = dim(y[[i]])[1], nrep = dim(y[[i]])[2], nspec = dim(y[[i]])[3],
         elev = site.covariates.scaled[[i]][,"elevation_median"],
         road = site.covariates.scaled[[i]][,"distance_to_road_median"],
         reserve = site.covariates.scaled[[i]][,"distance_to_nature_reserve_boundary"],
         numleeches = numleeches[[i]],
         R = matrix(c(5,0,0,1), ncol = 2), df = 3,
         zst = zst[[i]]
    )
    # copy over most data
        model.data[["model12b"]][[i]] <- model.data[["model12a"]][[i]]
        model.data[["model12c"]][[i]] <- model.data[["model12a"]][[i]]
    # add sampling covariates
        model.data[["model12b"]][[i]]$othertaxa <- othertaxa[[i]]
        model.data[["model12c"]][[i]]$humanfraction <- humanfraction[[i]]
}

model.data[["model18a"]] <- list()
model.data[["model18b"]] <- list()
model.data[["model18c"]] <- list()
for (i in names(leech.augmented)) {
    model.data[["model18a"]][[i]] <- list(y = y[[i]], nsite = dim(y[[i]])[1], nrep = dim(y[[i]])[2], nspec = dim(y[[i]])[3],
         road = site.covariates.scaled[[i]][,"distance_to_road_median"],
         reserve = site.covariates.scaled[[i]][,"distance_to_nature_reserve_boundary"],
         numleeches = numleeches[[i]],
         R = matrix(c(5,0,0,1), ncol = 2), df = 3,
         zst = zst[[i]]
    )
    # copy over most data
        model.data[["model18b"]][[i]] <- model.data[["model18a"]][[i]]
        model.data[["model18c"]][[i]] <- model.data[["model18a"]][[i]]
    # add sampling covariates
        model.data[["model18b"]][[i]]$othertaxa <- othertaxa[[i]]
        model.data[["model18c"]][[i]]$humanfraction <- humanfraction[[i]]
}

model.data[["model23a"]] <- list()
model.data[["model23b"]] <- list()
model.data[["model23c"]] <- list()
for (i in names(leech.augmented)) {
    model.data[["model23a"]][[i]] <- list(y = y[[i]], nsite = dim(y[[i]])[1], nrep = dim(y[[i]])[2], nspec = dim(y[[i]])[3],
         elev = site.covariates.scaled[[i]][,"elevation_median"],
         reserve = site.covariates.scaled[[i]][,"distance_to_nature_reserve_boundary"],
         numleeches = numleeches[[i]],
         R = matrix(c(5,0,0,1), ncol = 2), df = 3,
         zst = zst[[i]]
    )
    # copy over most data
        model.data[["model23b"]][[i]] <- model.data[["model23a"]][[i]]
        model.data[["model23c"]][[i]] <- model.data[["model23a"]][[i]]
    # add sampling covariates
        model.data[["model23b"]][[i]]$othertaxa <- othertaxa[[i]]
        model.data[["model23c"]][[i]]$humanfraction <- humanfraction[[i]]
}

rm(i)
```

```{r initial values}
# functions to set initial values for each model variant

# note that zst and nspec are evaluated at runtime, meaning that they will be the objects stored in the model data
inits <- list()

inits[["model01a"]] <- function() list(
    z = zst,
    omega = matrix(c(1,0,0,1), ncol = 2), eta = matrix(0, nrow = nspec, ncol = 2),
    betalpsi1 = rnorm(nspec), betalpsi2 = rnorm(nspec), betalpsi3 = rnorm(nspec), betalpsi4 = rnorm(nspec), betalpsi5 = rnorm(nspec),
    betalp1 = rnorm(nspec)
)
inits[["model01b"]] <- function() list(
    z = zst,
    omega = matrix(c(1,0,0,1), ncol = 2), eta = matrix(0, nrow = nspec, ncol = 2),
    betalpsi1 = rnorm(nspec), betalpsi2 = rnorm(nspec), betalpsi3 = rnorm(nspec), betalpsi4 = rnorm(nspec), betalpsi5 = rnorm(nspec),
    betalp1 = rnorm(nspec), betalp2 = rnorm(nspec)
)
inits[["model01c"]] <- inits[["model01b"]]

inits[["model12a"]] <- function() list(
    z = zst,
    omega = matrix(c(1,0,0,1), ncol = 2), eta = matrix(0, nrow = nspec, ncol = 2),
    betalpsi1 = rnorm(nspec), betalpsi3 = rnorm(nspec), betalpsi5 = rnorm(nspec),
    betalp1 = rnorm(nspec)
)
inits[["model12b"]] <- function() list(
    z = zst,
    omega = matrix(c(1,0,0,1), ncol = 2), eta = matrix(0, nrow = nspec, ncol = 2),
    betalpsi1 = rnorm(nspec), betalpsi3 = rnorm(nspec), betalpsi5 = rnorm(nspec),
    betalp1 = rnorm(nspec), betalp2 = rnorm(nspec)
)
inits[["model12c"]] <- inits[["model12b"]]

inits[["model23a"]] <- function() list(
    z = zst,
    omega = matrix(c(1,0,0,1), ncol = 2), eta = matrix(0, nrow = nspec, ncol = 2),
    betalpsi1 = rnorm(nspec), betalpsi5 = rnorm(nspec),
    betalp1 = rnorm(nspec)
)
inits[["model23b"]] <- function() list(
    z = zst,
    omega = matrix(c(1,0,0,1), ncol = 2), eta = matrix(0, nrow = nspec, ncol = 2),
    betalpsi1 = rnorm(nspec), betalpsi5 = rnorm(nspec),
    betalp1 = rnorm(nspec), betalp2 = rnorm(nspec)
)
inits[["model23c"]] <- inits[["model23b"]]

inits[["model18a"]] <- function() list(
    z = zst,
    omega = matrix(c(1,0,0,1), ncol = 2), eta = matrix(0, nrow = nspec, ncol = 2),
    betalpsi3 = rnorm(nspec), betalpsi5 = rnorm(nspec),
    betalp1 = rnorm(nspec)
)
inits[["model18b"]] <- function() list(
    z = zst,
    omega = matrix(c(1,0,0,1), ncol = 2), eta = matrix(0, nrow = nspec, ncol = 2),
    betalpsi3 = rnorm(nspec), betalpsi5 = rnorm(nspec),
    betalp1 = rnorm(nspec), betalp2 = rnorm(nspec)
)
inits[["model18c"]] <- inits[["model18b"]]
```

```{r parameters to monitor}
params <- list()

params[["model01a"]] <- c("mu.eta", "omega",
    "mu.betalpsi1", "sd.betalpsi1", "mu.betalpsi2", "sd.betalpsi2", "mu.betalpsi3", "sd.betalpsi3", "mu.betalpsi4", "sd.betalpsi4", "mu.betalpsi5", "sd.betalpsi5",
    "mu.betalp1", "sd.betalp1", "Nsite", "occ", "rho", "z", "psi", "lp0", "lpsi0")
params[["model01b"]] <- c("mu.eta", "omega",
    "mu.betalpsi1", "sd.betalpsi1", "mu.betalpsi2", "sd.betalpsi2", "mu.betalpsi3", "sd.betalpsi3", "mu.betalpsi4", "sd.betalpsi4", "mu.betalpsi5", "sd.betalpsi5",
    "mu.betalp1", "sd.betalp1", "mu.betalp2", "sd.betalp2", "Nsite", "occ", "rho", "z", "psi", "lp0", "lpsi0")
params[["model01c"]] <- params[["model01b"]]

params[["model12a"]] <- c("mu.eta", "omega",
    "mu.betalpsi1", "sd.betalpsi1", "mu.betalpsi3", "sd.betalpsi3", "mu.betalpsi5", "sd.betalpsi5",
    "mu.betalp1", "sd.betalp1", "Nsite", "occ", "rho", "z", "psi", "lp0", "lpsi0")
params[["model12b"]] <- c("mu.eta", "omega",
    "mu.betalpsi1", "sd.betalpsi1", "mu.betalpsi3", "sd.betalpsi3", "mu.betalpsi5", "sd.betalpsi5",
    "mu.betalp1", "sd.betalp1", "mu.betalp2", "sd.betalp2", "Nsite", "occ", "rho", "z", "psi", "lp0", "lpsi0")
params[["model12c"]] <- params[["model12b"]]

params[["model23a"]] <- c("mu.eta", "omega",
    "mu.betalpsi1", "sd.betalpsi1", "mu.betalpsi5", "sd.betalpsi5",
    "mu.betalp1", "sd.betalp1", "Nsite", "occ", "rho", "z", "psi", "lp0", "lpsi0")
params[["model23b"]] <- c("mu.eta", "omega",
    "mu.betalpsi1", "sd.betalpsi1", "mu.betalpsi5", "sd.betalpsi5",
    "mu.betalp1", "sd.betalp1", "mu.betalp2", "sd.betalp2", "Nsite", "occ", "rho", "z", "psi", "lp0", "lpsi0")
params[["model23c"]] <- params[["model23b"]]

params[["model18a"]] <- c("mu.eta", "omega",
    "mu.betalpsi3", "sd.betalpsi3", "mu.betalpsi5", "sd.betalpsi5",
    "mu.betalp1", "sd.betalp1", "Nsite", "occ", "rho", "z", "psi", "lp0", "lpsi0")
params[["model18b"]] <- c("mu.eta", "omega",
    "mu.betalpsi3", "sd.betalpsi3", "mu.betalpsi5", "sd.betalpsi5",
    "mu.betalp1", "sd.betalp1", "mu.betalp2", "sd.betalp2", "Nsite", "occ", "rho", "z", "psi", "lp0", "lpsi0")
params[["model18c"]] <- params[["model18b"]]
```

```{r save data to files}
models <- c("model01a", "model01b", "model01c",
            "model12a", "model12b", "model12c",
            "model23a", "model23b", "model23c",
            "model18a", "model18b", "model18c")
datasets <- c("LSU", "SSU")

for (m in models) {
    for (d in datasets) {
        jags.data <- list(
            model.data = model.data[[m]][[d]],
            inits = inits[[m]],
            params = params[[m]],
            mcmc.settings = mcmc.settings
        )
        jags.data.filename <- paste0("Ailaoshan_occupancy_MCMC_", m, "_", d, "_data.Rdata")
        save(jags.data, file=jags.data.filename)
    }
}

rm(models, datasets, jags.data.filename, jags.data, m, d)
```

```{r save prepared MCMC data}
# saves a copy of all prepared data
save(humanfraction, inits, leech, leech.supplement, leech.augmented, mcmc.settings, model.data, nsurvey, numleeches,
     obs.occ, othertaxa, params, site.covariates.scaled, y, zst, file = "Ailaoshan_occupancy_MCMC_prepare.Rdata")

```

```{r}

```

